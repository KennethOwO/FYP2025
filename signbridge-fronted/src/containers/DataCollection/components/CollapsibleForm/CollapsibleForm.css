.collapsible-container {
    margin-bottom: 10px;
    width: 85%;
}

.collapsible-content {
    position: relative;
    top: 50%;
    left: 50%;
    transform: translate(-50%, 0);
    width: 100%;
    max-width: 100%;
    margin-top: 10px;
    margin-bottom: 20px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    border-radius: 8px;
    transition: width 0.2s;
    background-color: white;
}

.collapsible-content-header {
    cursor: pointer;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 20px;
    border-radius: 8px 8px 0 0;
}

.collapsible-content.not-opened {
    width: 85%;
    cursor: pointer;
}

.collapsible-content.not-opened:hover {
    width: 100%;
    background-color: rgb(224, 224, 224);
}

.collapsible-content.opened {
    width: 100%;
    border: 1px solid #a3a3a3;
}

.collapsible-content-header.admin {
    background-color: #a4825e;
}

.collapsible-content-header.signexpert {
    background-color: rgb(198, 198, 198);
}

.collapsible-content-background {
    padding: 20px;
}

.collapsible-content-details {
    height: auto;
    justify-content: space-evenly;
    margin-bottom: 30px;
}

.collapsible-content-header .number {
    width: 20%;
}

.collapsible-content-header .dateTime {
    width: 35%;
}

.collapsible-content-header .status {
    width: 35%;
}

.collapsible-content-header .expand-icon {
    margin-left: auto;
}

.personal-details,
.sentence-details,
.content-left {
    flex: 1;
    margin: 0 10px;
}

.personal-details h2,
.sentence-details h2,
.content-left h2 {
    display: inline-block;
    color: #3a3a3a;
    font-size: 22px;
    margin-bottom: 15px;
}

.personal-details .ant-descriptions {
    margin-bottom: 20px;
    padding: 10px;
}

.personal-details .ant-descriptions-bordered {
    border: 1px solid #ccc;
    border-radius: 4px;
}

.personal-details .ant-descriptions-item-label {
    font-size: 20px;
    color: rgb(68, 68, 68) !important;
    /* box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.2); */
    background-color: #ececec !important;
    width: 25%;
    font-weight: 450;
}

/* .personal-details.admin .ant-descriptions-item-label {
    background-color: #FFF1C5 !important;
}

.personal-details.signexpert .ant-descriptions-item-label {
    background-color: #E7F0FD !important;
} */

.personal-details .ant-descriptions-item-content {
    font-size: 18px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.sentence-details-content p {
    border-radius: 5px;
    padding: 10px;
    background-color: #ececec;
    font-size: 18px;
    margin: 0 10px;
}

.collapsible-content-footer {
    display: flex;
    justify-content: space-between;
    align-items: flex-end;
    flex-wrap: wrap;
    width: 100%;
}

.collapsible-content-footer .content-left {
    flex-grow: 1;
    margin-top: -10px;
}

.collapsible-content-footer .content-left .ant-descriptions-bordered {
    border: 1px solid #ccc;
    border-radius: 4px;
}

.ant-descriptions.ant-descriptions-bordered .ant-descriptions-row {
    border-bottom: 1px solid #e8e8e8;
}

.collapsible-content-footer .content-left .ant-descriptions {
    max-width: 100%;
    padding: 10px;
}

.collapsible-content-footer .content-left .ant-descriptions-item-label {
    font-size: 20px;
    color: rgb(68, 68, 68) !important;
    background-color: #ececec !important;
    width: 25%;
    font-weight: 450;
}

.collapsible-content-footer .content-left .ant-descriptions-item-content {
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 16px;
    color: red;
    width: 100%;
    box-sizing: border-box;
    text-align: center;
}

.collapsible-content-footer .buttons-right {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-end;
    margin-top: 10px;
}

.dataset-button-container {
    display: flex;
    /* flex-wrap: wrap; */
    justify-content: flex-end;
}

.dataset-button-individual {
    margin-left: 10px;
}

@media (max-width: 480px) {
    .collapsible-container {
        width: 100%;
    }

    .collapsible-content-header {
        flex-direction: column;
        align-items: flex-start;
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        width: 100%;
        margin-bottom: 5px;
        font-size: 16px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    .buttons-right {
        align-items: center;
        justify-items: center;
    }

    .collapsible-content-footer .content-left {
        width: 100%;
        display: flex;
        flex-direction: column;
        margin: 0;
        padding: 0;
        text-align: left;
    }

    .collapsible-content-footer .buttons-right {
        width: 100%;
        justify-content: center;
        margin-top: 10px;
    }

    .dataset-button-container {
        display: flex;
        justify-content: center;
        gap: 10px;
    }

    .dataset-button-individual {
        width: 100%;
        margin-bottom: 10px;
    }

    .personal-details h2,
    .sentence-details h2,
    .content-left h2 {
        font-size: 18px;
    }

    .sentence-details-content p {
        font-size: 14px;
        padding: 8px;
    }

    .personal-details .ant-descriptions-item-content,
    .collapsible-content-footer .content-left .ant-descriptions-item-content {
        font-size: 12px;
    }

    .personal-details .ant-descriptions-item-label,
    .collapsible-content-footer .content-left .ant-descriptions-item-label {
        font-size: 14px;
    }

    .expand-icon {
        position: absolute;
        right: 20px;
        top: 20px;
        margin-top: 0;
    }
}

/* Devices between 481px and 640px (mobile landscape) */
@media (min-width: 481px) and (max-width: 640px) {
    .collapsible-container {
        width: 90%;
    }

    .collapsible-content-header {
        padding: 15px;
        flex-direction: column;
        align-items: flex-start;
    }

    .personal-details h2,
    .sentence-details h2,
    .content-left h2 {
        font-size: 20px;
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        font-size: 16px;
        margin-left: 10px;
    }

    .collapsible-content-header .number {
        width: 100%;
        font-size: 16px;
    }

    .collapsible-content-header .status {
        width: 100%;
        font-size: 16px;
    }

    .collapsible-content-header .dateTime {
        width: 100%;
        font-size: 16px;
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    .collapsible-content-footer {
        flex-direction: column;
        /* Stack elements vertically */
        align-items: center;
        /* Center elements horizontally */
    }

    .collapsible-content-footer .content-left {
        width: 100%;
        display: flex;
        flex-direction: column;
        margin: 0;
        padding: 0;
        text-align: left;
    }

    .collapsible-content-footer .buttons-right {
        width: 100%;
        justify-content: center;
        margin-top: 10px;
    }

    .dataset-button-container {
        display: flex;
        justify-content: center;
        gap: 10px;
    }

    .dataset-button-individual {
        margin-left: 10px;
    }

    .sentence-details-content p {
        font-size: 16px;
        padding: 8px;
    }

    .personal-details .ant-descriptions-item-content,
    .collapsible-content-footer .content-left .ant-descriptions-item-content {
        font-size: 14px;
    }

    .personal-details .ant-descriptions-item-label,
    .collapsible-content-footer .content-left .ant-descriptions-item-label {
        font-size: 16px;
    }

    .expand-icon {
        position: absolute;
        right: 20px;
        top: 20px;
        margin-top: 0;
    }
}

/* Devices between 641px and 768px (tablets and small devices) */
@media (min-width: 641px) and (max-width: 768px) {
    .collapsible-container {
        width: 85%;
    }

    .collapsible-content-header {
        padding: 18px;
    }

    .personal-details h2,
    .sentence-details h2,
    .content-left h2 {
        font-size: 20px;
    }

    .sentence-details-content p {
        font-size: 18px;
    }

    .collapsible-content-footer {
        flex-direction: column;
    }

    .collapsible-content-footer .content-left {
        width: 100%;
        display: flex;
        flex-direction: column;
        margin: 0;
        padding: 0;
        text-align: left;
    }

    .collapsible-content-footer .buttons-right {
        width: 100%;
        justify-content: center;
        margin-top: 10px;
    }

    .dataset-button-container {
        display: flex;
        justify-content: center;
        gap: 10px;
    }

    .dataset-button-individual {
        margin-left: 10px;
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        font-size: 17px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    .personal-details .ant-descriptions-item-content,
    .collapsible-content-footer .content-left .ant-descriptions-item-content {
        font-size: 14px;
    }

    .personal-details .ant-descriptions-item-label,
    .collapsible-content-footer .content-left .ant-descriptions-item-label {
        font-size: 16px;
    }

    .expand-icon {
        position: absolute;
        right: 20px;
        top: 20px;
        margin-top: 0;
    }
}

/* For devices with a width between 769px and 1024px (tablets) */
@media (min-width: 769px) and (max-width: 1024px) {
    .collapsible-container {
        width: 85%;
    }

    .collapsible-content-header {
        flex-direction: row;
        padding: 20px;
    }

    .collapsible-content-footer {
        flex-direction: column;
    }

    .collapsible-content-footer .content-left {
        width: 100%;
        display: flex;
        flex-direction: column;
        margin: 0;
        padding: 0;
        text-align: left;
    }

    .collapsible-content-footer .buttons-right {
        width: 100%;
        justify-content: center;
        margin-top: 10px;
    }

    .dataset-button-container {
        display: flex;
        justify-content: center;
        gap: 10px;
    }

    .dataset-button-individual {
        margin-left: 10px;
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        font-size: 20px;
        white-space: nowrap;
        /* Prevents text from wrapping */
        overflow: hidden;
        /* Hides any overflow text */
        text-overflow: ellipsis;
        /* Adds ellipsis when text overflows */
    }

    .expand-icon {
        position: absolute;
        right: 20px;
        top: 20px;
        margin-top: 0;
    }
}

@media (min-width: 1025px) {
    .collapsible-content.not-opened {
        width: 90%;
        cursor: pointer;
    }

    .collapsible-content.not-opened:hover {
        width: 100%;
        background-color: rgb(224, 224, 224);
    }

    .collapsible-content-header .number,
    .collapsible-content-header .status,
    .collapsible-content-header .dateTime {
        font-size: 20px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }
}