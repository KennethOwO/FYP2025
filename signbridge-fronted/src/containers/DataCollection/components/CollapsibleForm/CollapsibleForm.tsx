import React, { useState, useEffect } from "react";
import "./CollapsibleForm.css";
import { Button } from "../../../../components/Button/Button";
import { Descriptions } from "antd";
import {
  getDemoVidById,
  getAvatarVidById,
} from "../../../../services/dataset.service";
import { saveAs } from "file-saver";

import { toast } from "react-hot-toast";
import { CreateNotification } from "../../../../services/notification.service";
import VideoUpload from "../VideoUpload/VideoUpload";
import DownloadButton from "../../../../components/DownloadButton/DownloadButton";
import moment from "moment";
import { useTranslation } from "react-i18next";
import ButtonProcessing from "../../../../components/ButtonProcessing/ButtonProcessing";

import { useUserStore } from "@root/store/userStore";
import { getAuth } from "firebase/auth";

interface CollapsibleFormProps {
  number: string;
  form_id: number;
  dateTime: string;
  status_en: string;
  status_bm: string;
  name: string;
  email: string;
  text: string;
  video_link: string;
  avatar_link?: string;
  user?: string;
  user_id?: number;
  video_name: string;
  avatar_name: string;
  handleSubmit: Function;
  handleDelete: Function;
}

const CollapsibleForm: React.FC<CollapsibleFormProps> = ({
  number,
  form_id,
  dateTime,
  status_en,
  status_bm,
  name,
  email,
  text,
  video_link,
  avatar_link,
  user,
  user_id,
  video_name,
  avatar_name,
  handleSubmit,
  handleDelete,
}) => {
  const { t, i18n } = useTranslation();
  const currentSelectedLanguage = localStorage.getItem("i18nextLng") || "en";
  const { user: userStore } = useUserStore();
  const currentUser = getAuth().currentUser;

  const downloadVideo = async (type?: string) => {
    try {
      let video;
      if (type === "demo" && currentUser) {
        video = await getDemoVidById(form_id, currentUser); // Assuming getDemoVidById returns a Promise
        const blob = new Blob([video.data], { type: "video/mp4" }); // Assuming video.data contains blob data and video.contentType contains the content type

        saveAs(blob, video_name);
      } else if (type === "avatar" && currentUser) {
        video = await getAvatarVidById(form_id, currentUser); // Assuming getDemoVidById returns a Promise
        const blob = new Blob([video.data], { type: "video/mp4" }); // Assuming video.data contains blob data and video.contentType contains the content type

        saveAs(blob, avatar_name);
      }
    } catch (error) {
      console.error("Error downloading video:", error);
      // Handle download errors gracefully (e.g., display error message)
    }
  };

  const [isOpen, setIsOpen] = useState<boolean>(false);

  //Video Control
  const [videoInfo, setVideoInfo] = useState(null);
  const [uploadedVideo, setUploadedVideo] = useState<string | null>("");

  const toggleOpen = () => {
    if (isOpen) {
      setIsOpen(false);
    }
  };

  const closeForm = () => {
    if (!isOpen) {
      setIsOpen(true);
    }
  };

  const personal_details = [
    {
      key: "1",
      label: t("dc_name"),
      children: <span className="personal-details-info">{name}</span>,
    },
    {
      key: "2",
      label: t("dc_email"),
      children: <span className="personal-details-info">{email}</span>,
    },
  ];

  const video_details_with_video_upload = [
    {
      key: "1",
      label: t("dc_demo_video"),
      children: (
        <span className="video-details-info">
          <DownloadButton downloadVideo={downloadVideo} type="demo" />
        </span>
      ),
    },
    {
      key: "2",
      label: t("dc_avatar_video"),
      children: (
        <span className="video-details-info">
          <VideoUpload videoInfo={videoInfo} setVideoInfo={setVideoInfo} />
        </span>
      ),
    },
  ];

  const video_details_with_information = [
    {
      key: "1",
      label: t("dc_demo_video"),
      children: (
        <span className="video-details-info">
          <DownloadButton downloadVideo={downloadVideo} type="demo" />
        </span>
      ),
    },
    {
      key: "2",
      label: t("dc_avatar_video"),
      children:
        avatar_link !== "" ? (
          <span className="video-details-info">
            <DownloadButton downloadVideo={downloadVideo} type="avatar" />
          </span>
        ) : (
          <span className="video-details-info">{t("no_video_submitted")}</span>
        ),
    },
  ];

  // For notification
  const [isLoading, setIsLoading] = useState(false);

  // for public (accept/reject)
  const handleSEAcceptPublicButtonClick = async () => {
    try {
      const notificationData = {
        receiver_id: user_id,
        sender_id: 2,
        message_en: "has accepted your text.",
        message_bm: "telah menerima teks anda.",
        sign_text: text,
        status: 0,
        type: "Text Verification",
        type_value: "accepted",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
    try {
      const notificationData = {
        receiver_id: 1,
        sender_id: 2,
        message_en: "has assigned new text.",
        message_bm: "telah menetapkan teks baru.",
        sign_text: text,
        status: 0,
        type: "New Task",
        type_value: "newtask",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
  };

  const handleSERejectPublicButtonClick = async () => {
    try {
      const notificationData = {
        receiver_id: user_id,
        sender_id: 2,
        message_en: "has rejected your text.",
        message_bm: "telah menolak teks anda.",
        sign_text: text,
        status: 0,
        type: "Text Verification",
        type_value: "rejected",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
  };

  // for admin (accept/reject)
  const handleSEAcceptAdminButtonClick = async () => {
    try {
      const notificationData = {
        receiver_id: 1,
        sender_id: 2,
        message_en: "has accepted your avatar.",
        message_bm: "telah menerima avatar anda.",
        sign_text: text,
        status: 0,
        type: "Task Confirmation",
        type_value: "accepted",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
  };

  const handleSERejectAdminButtonClick = async () => {
    try {
      const notificationData = {
        receiver_id: 1,
        sender_id: 2,
        message_en: "has rejected your avatar.",
        message_bm: "telah menolak avatar anda.",
        sign_text: text,
        status: 0,
        type: "Task Confirmation",
        type_value: "rejected",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
  };
  // For admin upload
  const handleAdminButtonClick = async () => {
    try {
      const notificationData = {
        receiver_id: 2,
        sender_id: parseInt(userStore?.user_id?.toString() || "0"),
        message_en: "has uploaded the avatar.",
        message_bm: "telah memuat naik avatar.",
        sign_text: text,
        status: 0,
        type: "Waiting for Verification",
        type_value: "waitingforverification",
        created_at: new Date().toISOString(),
      };
      if (currentUser) {
        await CreateNotification(notificationData, currentUser);
        toast.success(t("notifSuccess"));
      }
    } catch (error) {
      console.error("Error sending notification:", error);
      toast.error(t("notifFailed"));
    }
  };
  return (
    <div className="collapsible-container ">
      <div
        className={`collapsible-content ${isOpen ? "opened" : "not-opened"}`}
        onClick={closeForm}
        data-testid={`collapsible-content_${form_id}`}
      >
        <div
          className={`collapsible-content-header ${user}`}
          onClick={toggleOpen}
        >
          <h2 className="number">No: {number}</h2>
          <h2 className="status">
            Status: {currentSelectedLanguage === "en" ? status_en : status_bm}
          </h2>
          <h2 className="dateTime">
            {t("dc_date")}: {moment(dateTime).format("YYYY-MM-DD HH:mm:ss")}
          </h2>
          <div className="expand-icon">
            {isOpen ? <span>&#9650;</span> : <span>&#9660;</span>}
          </div>
        </div>
        {isOpen && (
          <div className="collapsible-content-background">
            <div className="collapsible-content-details">
              <div className={`personal-details ${user}`}>
                <h2>{t("dc_personal_details")}</h2>
                <Descriptions items={personal_details} bordered column={1} />
              </div>
              <div className="sentence-details">
                <h2>{t("dc_text_sentence")}</h2>
                <div className="sentence-details-content">
                  <p>{text}</p>
                </div>
              </div>
            </div>
            <div className="collapsible-content-footer">
              <div className={`content-left ${user}`}>
                <h2>{t("dc_video_details")}</h2>
                {user === "admin" && (
                  <>
                    {status_en === "Rejected" && status_bm === "Ditolak" ? (
                      <Descriptions
                        items={video_details_with_video_upload}
                        bordered
                        column={1}
                      />
                    ) : status_en === "In Progress" &&
                      status_bm === "Dalam Proses" ? (
                      <Descriptions
                        items={video_details_with_video_upload}
                        bordered
                        column={1}
                      />
                    ) : (
                      <Descriptions
                        items={video_details_with_information}
                        bordered
                        column={1}
                      />
                    )}
                  </>
                )}

                {user === "signexpert" && (
                  <Descriptions
                    items={video_details_with_information}
                    bordered
                    column={1}
                  />
                )}
              </div>

              <div className="buttons-right">
                {user === "signexpert" && (
                  <>
                    {status_en === "New" && status_bm === "Baru" && (
                      <div className="dataset-button-container">
                        <div className="dataset-button-individual">
                          <Button
                            type="button"
                            onClick={() => {
                              const updateData = {
                                status_SE_en: "Awaiting Accept",
                                status_SE_bm: "Menunggu Penerimaan",
                                status_Admin_en: "New",
                                status_Admin_bm: "Baru",
                              };
                              const updatedMessage = "New_Request_Accepted";

                              handleSubmit(
                                form_id,
                                updateData,
                                false,
                                updatedMessage
                              );
                              handleSEAcceptPublicButtonClick();
                            }}
                            buttonStyle="btn--accept" // Style for accept button
                            buttonSize="btn--large"
                          >
                            {t("dc_accept")}
                          </Button>
                        </div>
                        <div className="dataset-button-individual">
                          <Button
                            type="button"
                            onClick={() => {
                              const updatedMessage = "Request_Cancelled";
                              handleDelete(form_id, updatedMessage);
                              handleSERejectPublicButtonClick();
                            }}
                            buttonStyle="btn--cancel" // Style for cancel button
                            buttonSize="btn--large"
                          >
                            {t("dc_cancel")}
                          </Button>
                        </div>
                      </div>
                    )}
                    {status_en === "Awaiting Verification" &&
                      status_bm === "Menunggu Pengesahan" && (
                        <div className="dataset-button-container">
                          <div className="dataset-button-individual">
                            <Button
                              type="button"
                              onClick={() => {
                                const updateData = {
                                  status_SE_en: "Verified",
                                  status_SE_bm: "Disahkan",
                                  status_Admin_en: "Verified",
                                  status_Admin_bm: "Disahkan",
                                };
                                const updatedMessage = "Request_Verified";
                                handleSubmit(
                                  form_id,
                                  updateData,
                                  false,
                                  updatedMessage
                                );
                                handleSEAcceptAdminButtonClick();
                              }}
                              buttonStyle="btn--send"
                              buttonSize="btn--large"
                            >
                              {t("dc_verify")}
                            </Button>
                          </div>
                          <div className="dataset-button-individual">
                            <Button
                              type="button"
                              onClick={() => {
                                const updateData = {
                                  status_SE_en: "Rejected",
                                  status_SE_bm: "Ditolak",
                                  status_Admin_en: "Rejected",
                                  status_Admin_bm: "Ditolak",
                                };
                                const updatedMessage = "Request_Rejected";
                                handleSubmit(
                                  form_id,
                                  updateData,
                                  false,
                                  updatedMessage
                                );
                                handleSERejectAdminButtonClick();
                              }}
                              buttonStyle="btn--cancel"
                              buttonSize="btn--large"
                            >
                              {t("dc_reject")}
                            </Button>
                          </div>
                        </div>
                      )}
                  </>
                )}
                {user === "admin" && (
                  <>
                    {status_en === "New" && status_bm === "Baru" && (
                      <Button
                        type="button"
                        onClick={() => {
                          const updateData = {
                            status_SE_en: "In Progress",
                            status_SE_bm: "Dalam Proses",
                            status_Admin_en: "In Progress",
                            status_Admin_bm: "Dalam Proses",
                          };
                          const updatedMessage = "Request_Accepted_by_Admin";
                          handleSubmit(
                            form_id,
                            updateData,
                            false,
                            updatedMessage
                          );
                        }}
                        buttonStyle="btn--accept" // Style for accept button
                        buttonSize="btn--large"
                      >
                        {t("dc_accept")}
                      </Button>
                    )}
                    {status_en === "In Progress" &&
                      status_bm === "Dalam Proses" && (
                        <ButtonProcessing
                          onClick={async () => {
                            if (videoInfo) {
                              setIsLoading(true);

                              const updateData = {
                                status_SE_en: "Awaiting Verification",
                                status_SE_bm: "Menunggu Pengesahan",
                                status_Admin_en: "Awaiting Verification",
                                status_Admin_bm: "Menunggu Pengesahan",
                              };
                              const updatedMessage = "Request_Updated_by_Admin";
                              try {
                                const response = await handleSubmit(
                                  form_id,
                                  updateData,
                                  videoInfo,
                                  updatedMessage
                                );

                                if (response) {
                                  handleAdminButtonClick();
                                }
                              } catch (error) {
                                console.error(
                                  "Error during handleSubmit:",
                                  error
                                );
                              } finally {
                                setIsLoading(false);
                              }
                            } else {
                              toast.error(t("mustUploadAvatar"));
                            }
                          }}
                          buttonStyle="btn--accept"
                          buttonSize="btn--large"
                          isLoading={isLoading}
                          setIsLoading={setIsLoading}
                        >
                          {t("dc_submit")}
                        </ButtonProcessing>
                      )}
                    {status_en === "Rejected" && status_bm === "Ditolak" && (
                      <ButtonProcessing
                        onClick={async () => {
                          if (videoInfo) {
                            setIsLoading(true);
                            const updateData = {
                              status_SE_en: "Awaiting Verification",
                              status_SE_bm: "Menunggu Pengesahan",
                              status_Admin_en: "Awaiting Verification",
                              status_Admin_bm: "Menunggu Pengesahan",
                            };
                            const updatedMessage = "Request_Updated_by_Admin";
                            try {
                              const response = await handleSubmit(
                                form_id,
                                updateData,
                                videoInfo,
                                updatedMessage
                              );

                              if (response) {
                                handleAdminButtonClick();
                              }
                            } catch (error) {
                              console.error(
                                "Error during handleSubmit:",
                                error
                              );
                            } finally {
                              setIsLoading(false);
                            }
                          } else {
                            toast.error(t("mustUploadAvatar"));
                          }
                        }}
                        buttonStyle="btn--accept"
                        buttonSize="btn--large"
                        isLoading={isLoading}
                        setIsLoading={setIsLoading}
                      >
                        {t("dc_submit")}
                      </ButtonProcessing>
                    )}
                  </>
                )}
              </div>
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default CollapsibleForm;
