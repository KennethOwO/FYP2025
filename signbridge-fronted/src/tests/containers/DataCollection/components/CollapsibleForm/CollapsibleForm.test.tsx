//Import necessary libraries -> either "@testing-library/react" or "react-test-renderer"
import { render, fireEvent } from "@testing-library/react";
import CollapsibleForm from "../../../../../containers/DataCollection/components/CollapsibleForm/CollapsibleForm";
import { cleanup } from "@testing-library/react";
jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key, // Mock the translation function to return the key
    i18n: {
      changeLanguage: jest.fn(), // Mock the changeLanguage function
    },
  }),
}));
jest.mock("moment", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    format: jest.fn(() => "2024-05-24 12:00:00"), // Mock the format method
  })),
}));
describe("Test Collapsible Form", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  afterEach(() => {
    // Clean up the DOM
    cleanup();
    // Reset mocks to their initial state
    jest.resetAllMocks();
  });
  const props = {
    number: "1",
    form_id: 1,
    dateTime: "2023-11-11",
    status_en: "New",
    status_bm: "Baru",
    name: "Luffy Lee",
    email: "luffynika@example.com",
    text: "Aku mahu makan nasi",
    video_link: "testing123.mp4",
    avatar_link: "",
    user: "admin",
    user_id: 1,
    video_name: "testing123",
    avatar_name: "testing123",
    handleSubmit: jest.fn(),
    handleDelete: jest.fn(),
  };
  it("should render correctly", () => {
    const { container, getByText } = render(<CollapsibleForm {...props} />);
    const header = getByText(`No: ${props.number}`).closest(
      ".collapsible-content-header"
    );

    if (header) {
      fireEvent.click(header);
    }

    expect(container).toMatchSnapshot();
  });
});
