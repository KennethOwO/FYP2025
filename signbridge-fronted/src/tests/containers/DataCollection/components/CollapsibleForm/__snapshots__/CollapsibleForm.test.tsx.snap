// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Collapsible Form should render correctly 1`] = `
<div>
  <div
    class="collapsible-container "
  >
    <div
      class="collapsible-content opened"
      data-testid="collapsible-content_1"
    >
      <div
        class="collapsible-content-header admin"
      >
        <h2
          class="number"
        >
          No: 
          1
        </h2>
        <h2
          class="status"
        >
          Status: 
          New
        </h2>
        <h2
          class="dateTime"
        >
          dc_date
          : 
          2024-05-24 12:00:00
        </h2>
        <div
          class="expand-icon"
        >
          <span>
            â–²
          </span>
        </div>
      </div>
      <div
        class="collapsible-content-background"
      >
        <div
          class="collapsible-content-details"
        >
          <div
            class="personal-details admin"
          >
            <h2>
              dc_personal_details
            </h2>
            <div
              class="ant-descriptions ant-descriptions-bordered css-dev-only-do-not-override-42nv3w"
            >
              <div
                class="ant-descriptions-view"
              >
                <table>
                  <tbody>
                    <tr
                      class="ant-descriptions-row"
                    >
                      <th
                        class="ant-descriptions-item-label"
                        colspan="1"
                      >
                        <span>
                          dc_name
                        </span>
                      </th>
                      <td
                        class="ant-descriptions-item-content"
                        colspan="1"
                      >
                        <span>
                          <span
                            class="personal-details-info"
                          >
                            Luffy Lee
                          </span>
                        </span>
                      </td>
                    </tr>
                    <tr
                      class="ant-descriptions-row"
                    >
                      <th
                        class="ant-descriptions-item-label"
                        colspan="1"
                      >
                        <span>
                          dc_email
                        </span>
                      </th>
                      <td
                        class="ant-descriptions-item-content"
                        colspan="1"
                      >
                        <span>
                          <span
                            class="personal-details-info"
                          >
                            luffynika@example.com
                          </span>
                        </span>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
          <div
            class="sentence-details"
          >
            <h2>
              dc_text_sentence
            </h2>
            <div
              class="sentence-details-content"
            >
              <p>
                Aku mahu makan nasi
              </p>
            </div>
          </div>
        </div>
        <div
          class="collapsible-content-footer"
        >
          <div
            class="content-left admin"
          >
            <h2>
              dc_video_details
            </h2>
            <div
              class="row"
            >
              <div
                class="col-md-6"
              >
                <div
                  class="ant-descriptions ant-descriptions-bordered css-dev-only-do-not-override-42nv3w"
                >
                  <div
                    class="ant-descriptions-view"
                  >
                    <table>
                      <tbody>
                        <tr
                          class="ant-descriptions-row"
                        >
                          <th
                            class="ant-descriptions-item-label"
                            colspan="1"
                          >
                            <span>
                              dc_demo_video
                            </span>
                          </th>
                          <td
                            class="ant-descriptions-item-content"
                            colspan="1"
                          >
                            <span>
                              <span
                                class="video-details-info"
                              >
                                <span
                                  class="video-details-info"
                                >
                                  <button
                                    class="ant-btn css-dev-only-do-not-override-42nv3w ant-btn-default"
                                    style="width: 175px; height: 40px;"
                                    type="button"
                                  >
                                    <span>
                                      <span
                                        aria-label="download"
                                        class="anticon anticon-download"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="download"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="64 64 896 896"
                                          width="1em"
                                        >
                                          <path
                                            d="M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                    <span
                                      style="margin-left: 8px;"
                                    >
                                      download_btn
                                    </span>
                                  </button>
                                </span>
                              </span>
                            </span>
                          </td>
                        </tr>
                        <tr
                          class="ant-descriptions-row"
                        >
                          <th
                            class="ant-descriptions-item-label"
                            colspan="1"
                          >
                            <span>
                              dc_avatar_video
                            </span>
                          </th>
                          <td
                            class="ant-descriptions-item-content"
                            colspan="1"
                          >
                            <span>
                              <span
                                class="video-details-info"
                              >
                                no_video_submitted
                              </span>
                            </span>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="buttons-right"
          >
            <button
              class="btn btn--accept btn--large"
              data-testid="button-btn--accept"
              type="button"
            >
              dc_accept
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
